### PR TITLE
Feature fast decimal decimal to int casts

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

**Comment by sopel39** [comment]
```
After:

Benchmark                                                         (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.decimalCastBenchmark  cast(l_38_30 as decimal(8, 0))  thrpt   40  0,179 ± 0,004  ops/ms
BenchmarkDecimalOperators.decimalCastBenchmark  cast(l_26_18 as decimal(8, 0))  thrpt   40  0,273 ± 0,005  ops/ms
BenchmarkDecimalOperators.decimalCastBenchmark  cast(l_20_12 as decimal(8, 0))  thrpt   40  0,679 ± 0,017  ops/ms
BenchmarkDecimalOperators.decimalCastBenchmark   cast(l_20_8 as decimal(8, 0))  thrpt   40  0,663 ± 0,027  ops/ms

Before:

# Run complete. Total time: 00:05:44

Benchmark                                                         (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.decimalCastBenchmark  cast(l_38_30 as decimal(8, 0))  thrpt   40  0,113 ± 0,003  ops/ms
BenchmarkDecimalOperators.decimalCastBenchmark  cast(l_26_18 as decimal(8, 0))  thrpt   40  0,109 ± 0,013  ops/ms
BenchmarkDecimalOperators.decimalCastBenchmark  cast(l_20_12 as decimal(8, 0))  thrpt   40  0,287 ± 0,010  ops/ms
BenchmarkDecimalOperators.decimalCastBenchmark   cast(l_20_8 as decimal(8, 0))  thrpt   40  0,313 ± 0,011  ops/ms
```

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 1]
```diff
@@ -411,6 +404,7 @@ public static double shortDecimalToDouble(long decimal, long precision, long sca
     @UsedByGeneratedCode
     public static double longDecimalToDouble(Slice decimal, long precision, long scale, long tenToScale)
     {
+        // TODO: optimize
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:407]: What to optimize. Either provide some explanation or drop note.

File: presto-main/src/main/java/com/facebook/presto/type/DecimalToDecimalCasts.java [hunk 2]
```diff
@@ -118,44 +119,39 @@ public static Slice shortToLongCast(
             long resultPrecision,
             long resultScale)
     {
-        return encodeUnscaledValue(bigintToBigintCast(BigInteger.valueOf(value), sourcePrecision, sourceScale, resultPrecision, resultScale));
+        return longToLongCast(unscaledDecimal(value), sourcePrecision, sourceScale, resultPrecision, resultScale);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalToDecimalCasts.java:122]: extra unnecessaery intermediate `Slice` instance is used in this flow.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 3]
```diff
@@ -23,6 +23,7 @@
 import java.nio.ByteOrder;
 
 import static com.facebook.presto.spi.type.Decimals.MAX_PRECISION;
+import static com.facebook.presto.spi.type.Decimals.longTenToNth;
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:26]: Provide some comment in commit message on nature of improvements.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 4]
```diff
@@ -44,7 +45,7 @@
     private static final int SIGN_INT_MASK = 1 << 31;
     private static final int SIGN_BYTE_MASK = 1 << 7;
     private static final int NUMBER_OF_LONGS = 2;
-    private static final int INT_MASK = 0xFFFFFFFF;
+    private static final long FULLBITS_64 = 0xFFFFFFFFFFFFFFFFL;
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:48]: ALL_BITS_SET_64

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 5]
```diff
@@ -59,6 +60,19 @@
...
+     */
+    private static final int MAX_POWER_TEN_INT = 9;
+    /**
+     * 10^18 fits in 2^63.
+     */
+    private static final int MAX_POWER_TEN_LONG = 18;
+    /**
+     * 10^x. All unsigned values.
+     */
+    private static final int[] POWER_TENS_INT = new int[MAX_POWER_TEN_INT + 1];
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:74]: POWERS_OF_TEN_INT

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 6]
```diff
@@ -59,6 +60,19 @@
      */
     private static final int[] POWER_FIVES_INT = new int[MAX_POWER_FIVE_INT + 1];
 
+    /**
+     * 10^9 fits in 2^31.
+     */
+    private static final int MAX_POWER_TEN_INT = 9;
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:66]: MAX_POWER_OF_TEN_INT

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 7]
```diff
@@ -59,6 +60,19 @@
...
+    /**
+     * 10^18 fits in 2^63.
+     */
+    private static final int MAX_POWER_TEN_LONG = 18;
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:70]: MAX_POWER_OF_TEN_LONG

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 8]
```diff
@@ -162,6 +181,13 @@ public static long unscaledDecimalToUnscaledLong(Slice decimal)
         return negative ? -lo : lo;
     }
 
+    public static long unscaledDecimalToUnscaledLongUnsafe(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:186]: s/lo/low

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:407]: removed

File: presto-main/src/main/java/com/facebook/presto/type/DecimalToDecimalCasts.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalToDecimalCasts.java:122]: The problem is that if we want to have reusable arithmetic we have to have some value storage (Slice in this case). That is why I opt for `result` parameter in arithmetic. We can also use accumulator `result` field as in ArrayConcatFunction to avoid Slice allocation

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 4 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:48]: done

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 5 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:74]: done

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 6 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:66]: done

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 7 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:70]: done

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 8 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:186]: let's make it separate patch to make it consistent after our PR are merged into dev branch. Let's us low/high from now

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 9]
```diff
@@ -373,6 +422,31 @@ private static void scaleDownFive(Slice decimal, int fiveScale, Slice result)
...
     }
 
+    /**
+     * Scale down the value for 10**tenScale (this := this / 5**tenScale). This
+     * method rounds-up, eg 44/10=4, 44/5=5.
+     */
+    private static void scaleDownTenRoundUp(Slice decimal, int tenScale, Slice result)
+    {
+        boolean round;
+        while (true) {
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:432]: make it `while (tenScale > 0)` and remove `break` from the loop.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 10]
```diff
@@ -373,6 +422,31 @@ private static void scaleDownFive(Slice decimal, int fiveScale, Slice result)
         }
     }
 
+    /**
+     * Scale down the value for 10**tenScale (this := this / 5**tenScale). This
+     * method rounds-up, eg 44/10=4, 44/5=5.
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:427]: WTF `44/10=4, 44/5=5`

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 11]
```diff
@@ -487,7 +531,20 @@ static void divide(Slice decimal, int divisor, Slice result)
...
+
+    private static void incrementUnsafe(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        if (lo != FULLBITS_64) {
+            setRawLong(decimal, 0, lo + 1);
+            return;
+        }
+
+        long hi = getLong(decimal, 1);
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:546]: check overflow?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 12]
```diff
@@ -500,6 +557,11 @@ private static void setNegative(Slice decimal, int v3, boolean negative)
         setRawInt(decimal, SIGN_INT_INDEX, v3 | (negative ? SIGN_INT_MASK : 0));
     }
 
+    private static void setNegative(Slice decimal, long hi, boolean negative)
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:560]: Having two methods:

`setNegative(Slice, long, boolean)` and `setNegative(Slice, int, boolean)` is very error prone. Use different names for both.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 9 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:432]: using do while

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 10 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:427]: :)

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 11 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:546]: it is unsafe version, no overflow check

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 12 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:560]: done
